### DISCUSSION TITLE
Standalone output in pnpm monorepo has invalid symlinks in node_modules

### DISCUSSION THREAD

**Discussion by revmischa** [vercel/next.js#40482] [Help] [answered]
I'm using next 12.3.0 with output: 'standalone' and pnpm. The `node_modules` directory in `standalone/` contains invalid symlinks, presumably because it's copying the symlinks from the project into the output dir.

![image](https://user-images.githubusercontent.com/245131/189251500-124e592b-078d-4aac-8648-a1fbb74c049d.png)

Perhaps this could be fixed by following symlinks when copying the node_modules to the output?

There is a reproduction repo here: https://github.com/jetbridge/next-pnpm-monorepo-repro

_Originally posted by @revmischa in https://github.com/vercel/next.js/issues/40150#issuecomment-1241389692_

**Comment by remorses** [comment, marked as answer]
I think you have to use `outputFileTracingRoot` if you are using a monorepo (the path points to the root of the monorepo)

```js
/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  swcMinify: true,
  output: "standalone",
  experimental: { outputFileTracingRoot: path.join(__dirname, '../'), }
};

module.exports = nextConfig;

```

**Reply by revmischa** [reply]
I'm making some progress but now I have this error:
```
    "Runtime.ImportModuleError: Error: Cannot find module '@swc/helpers/lib/_async_to_generator.js'",
    "Require stack:",
    "- /var/task/web/node_modules/next/dist/shared/lib/utils.js",
    "- /var/task/web/node_modules/next/dist/server/next-server.js",
    "- /var/task/web/server.js",
    "- /var/runtime/index.mjs",
```

**Reply by michrome** [reply]
Hey @revmischa were you able to root cause the `Cannot find module '@swc/helpers/lib/_async_to_generator.js'` error?

**Reply by revmischa** [reply]
If I remember, the issue was that I wasn't preserving symlinks in my function bundle.

If you're trying to deploy nextjs on AWS I suggest taking a look at https://github.com/jetbridge/cdk-nextjs - we've been dealing with a lot of these funny details and made an open source repo so that others do not have to struggle as much

**Reply by michrome** [reply]
Thanks for sharing - much appreciated.

**Reply by qhantom** [reply]
@revmischa I ran into the same error message that tells me 

```
Error: Cannot find module '@swc/helpers/_/_interop_require_default'
Require stack:
- /app/packages/app/node_modules/next/dist/shared/lib/constants.js
- /app/packages/app/node_modules/next/dist/server/config.js
- /app/packages/app/node_modules/next/dist/server/lib/router-server.js
- /app/packages/app/node_modules/next/dist/server/lib/start-server.js
- /app/packages/app/server.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
    at Module._load (node:internal/modules/cjs/loader:975:27)
    at Module.require (node:internal/modules/cjs/loader:1225:19)
    at require (node:internal/modules/helpers:177:18)
    at Object.<anonymous> (/app/packages/app/node_modules/next/dist/shared/lib/constants.js:271:34)
    at Module._compile (node:internal/modules/cjs/loader:1356:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32)
    at Module._load (node:internal/modules/cjs/loader:1013:12)
    at Module.require (node:internal/modules/cjs/loader:1225:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/app/packages/app/node_modules/next/dist/shared/lib/constants.js',
    '/app/packages/app/node_modules/next/dist/server/config.js',
    '/app/packages/app/node_modules/next/dist/server/lib/router-server.js',
    '/app/packages/app/node_modules/next/dist/server/lib/start-server.js',
    '/app/packages/app/server.js'
  ]
}
```

If I run the Dockerfile locally, it all works fine. But when I let the GitHub Action `docker/build-push-action@v5` build the image with no extra config, the Docker image is corrupted with the error above. How did you exactly resolve the issue regarding the symlink resolution? Thanks a lot!

**Reply by revmischa** [reply]
I'm now using https://open-next.js.org/ which is partly based on the above cdk-nextjs. You can look at its code to see.

**Comment by siuming-qiu** [comment]
<img width="2258" height="2124" alt="image" src="https://github.com/user-attachments/assets/9e3d0fdf-d43a-46ff-b5d0-540af6faadc8" />
still not work, anyone can help me ?...
